primitives/trie/src/cache/shared_cache.rs

```diff
@@ -0,0 +1,632 @@
+	pub(super) lru: NoHashingLruCache<ValueCacheKey<'static, H>, CachedValue<H>>,
+	/// The size of [`Self::lru`] in bytes.
+	pub(super) size_in_bytes: usize,
+	/// The maximum cache size of [`Self::lru`].
+	maximum_cache_size: CacheSize,
+	/// All known storage keys that are stored in [`Self::lru`].
+	///
+	/// This is used to de-duplicate keys in memory that use the
+	/// same [`SharedValueCache::storage_key`], but have a different
+	/// [`SharedValueCache::storage_root`].
+	known_storage_keys: HashSet<Arc<[u8]>>,
+}
+
+impl<H: Eq + std::hash::Hash + Clone + Copy + AsRef<[u8]>> SharedValueCache<H> {
+	/// Create a new instance.
+	fn new(cache_size: CacheSize) -> Self {
+		Self {
+			lru: NoHashingLruCache::unbounded_with_hasher(Default::default()),
+			size_in_bytes: 0,
+			maximum_cache_size: cache_size,
+			known_storage_keys: Default::default(),
+		}
+	}
+
+	/// Get the [`CachedValue`] for `key`.
+	///
+	/// This doesn't change the least recently order in the internal [`LruCache`].
+	pub fn get<'a>(&'a self, key: &ValueCacheKey<H>) -> Option<&'a CachedValue<H>> {
+		debug_assert!(
+			!matches!(key, ValueCacheKey::Hash { .. }),
+			"`get` can not be called with `Hash` variant as this may returns the wrong value."
+		);
+
+		self.lru.peek(unsafe {
+			// SAFETY
+			//
+			// We need to convert the lifetime to make the compiler happy. However, as
+			// we only use the `key` to looking up the value this lifetime conversion is
+			// safe.
+			mem::transmute::<&ValueCacheKey<'_, H>, &ValueCacheKey<'static, H>>(key)
+		})
+	}
+
+	/// Update the cache with the `added` values and the `accessed` values.
+	///
+	/// The `added` values are the ones that have been collected by doing operations on the trie and
+	/// now should be stored in the shared cache. The `accessed` values are only referenced by the
+	/// [`ValueCacheKeyHash`] and represent the values that were retrieved from this shared cache
+	/// through [`Self::get`]. These `accessed` values are being put to the front of the internal
+	/// [`LruCache`] like the `added` ones.
+	///
+	/// After the internal [`LruCache`] was updated, it is ensured that the internal [`LruCache`] is
+	/// inside its bounds ([`Self::maximum_size_in_bytes`]).
+	pub fn update(
+		&mut self,
+		added: impl IntoIterator<Item = (ValueCacheKey<'static, H>, CachedValue<H>)>,
+		accessed: impl IntoIterator<Item = ValueCacheKeyHash>,
+	) {
+		// The base size in memory per ([`ValueCacheKey<H>`], [`CachedValue`]).
+		let base_size = mem::size_of::<ValueCacheKey<H>>() + mem::size_of::<CachedValue<H>>();
+		let known_keys_entry_size = mem::size_of::<Arc<[u8]>>();
+
+		let update_size_in_bytes =
+			|size_in_bytes: &mut usize, r_key: Arc<[u8]>, known_keys: &mut HashSet<Arc<[u8]>>| {
+				// If the `strong_count == 2`, it means this is the last instance of the key.
+				// One being `r_key` and the other being stored in `keys`.
+				let last_instance = Arc::strong_count(&r_key) == 2;
```

Comment:
This looks fragile. Is it possible that some other thread is changing these references while this code is executing?

Reply:
This function is only callable when someone has write access to the `SharedValueCache`. Aka there can not be any other thread that is modifying this. And even if this would happen and we remove the latest instance here. What happen would be that there is some key that isn't tracked.